src/annotated.rs

```diff
@@ -0,0 +1,283 @@
+//! Utilities for extracting YAML with certain metadata.
+
+pub mod marked_yaml;
+
+use std::ops::{Index, IndexMut};
+
+use hashlink::LinkedHashMap;
+
+use crate::loader::parse_f64;
+
+/// YAML data for nodes that will contain annotations.
+///
+/// If you want a YAML node without annotations, see [`Yaml`].
+/// If you want a YAML node with annotations, see types using [`YamlData`] such as [`MarkedYaml`]
+///
+/// Unlike [`Yaml`] which only supports storing data, [`YamlData`] allows storing metadata
+/// alongside the YAML data. It is unlikely one would build it directly; it is mostly intended to
+/// be used, for instance, when parsing a YAML where retrieving markers / comments is relevant.
+///
+/// This definition is recursive. Each annotated node will be a structure storing the annotations
+/// and the YAML data. We need to have a distinct enumeration from [`Yaml`] because the type for
+/// the `Array` and `Hash` variants is dependant on that structure.
+///
+/// If we had written [`YamlData`] as:
+/// ```ignore
+/// pub enum YamlData {
+///   // ...
+///   Array(Vec<Yaml>),
+///   Hash(LinkedHashMap<Yaml, Yaml>),
+///   // ...
+/// }
+/// ```
+/// we would have stored metadata for the root node only. All subsequent nodes would be [`Yaml`],
+/// which does not contain any annotation.
+///
+/// Notable differences with [`Yaml`]:
+///   * Indexing cannot return `BadValue` and will panic instead.
```

Comment:
I doesn't have to be in this PR specifically - I'm happy to get this version for now, and leave this for future improvement. But I wonder if you couldn't just add a trait for that? `trait BadValue { fn bad_value() -> Self }` and add this constraint to `Node`. If it's important, that is.

Reply:
It turns out this is more difficult to implement than I thought since I'd need to initialize the `Marker` in `MarkedYaml`, but the fields are private.

I'll leave it as-is, but other than that, the function can be added to `LoadableYamlNode`.